### PR TITLE
Feat/recurring buy

### PR DESCRIPTION
**Pull request by prwelber**
## Description
Recurring Buy

## Change Type
Please enter one or more of the following: 
- Feature

## Testing Steps
This should be tested on testnet... `yarn start:testnet`

## Code Checklist
- [x] Code compiles successfully (verified via `yarn start`)
- [x] No lint issues exist (verified via `yarn lint`)
- [x] New and existing unit tests pass (verified via `yarn test`)
- [x] `README.md` and other documentation is updated as needed

### CODE REVIEW

File: packages/blockchain-wallet-v4-frontend/src/data/modules/coinify/reducers.js [hunk 1]
```diff
@@ -73,6 +77,29 @@ const coinify = (state = INITIAL_STATE, action) => {
...
+    case AT.COINIFY_SET_RECURRING_TRADE_END_TIME: {
+      return assocPath(['subscriptionData', 'endTime'], payload, state)
+    }
+    case AT.COINIFY_RESET_RECURRING_BUY: {
+      return mergeAll([
+        state,
+        { showRecurringModal: INITIAL_STATE.showRecurringModal },
+        { subscription: INITIAL_STATE.subscription },
+        { subscriptionData: INITIAL_STATE.subscriptionData }
+      ], state)
```
Comment (tony-blockchain) [inline packages/blockchain-wallet-v4-frontend/src/data/modules/coinify/reducers.js]: looks like an unneeded `state` param

File: packages/blockchain-wallet-v4-frontend/src/data/modules/coinify/sagas.js [hunk 2]
```diff
@@ -539,6 +562,76 @@ export default ({ coreSagas, networks }) => {
...
 
+  const recurringCheckoutInitialized = function * () {
+    const initialValues = { frequency: 'weekly', duration: null }
+    yield put(initialize('coinifyRecurringCheckout', initialValues))
+  }
+
+  const handleRecurringFormChange = function * (action) {
+    try {
+      const form = path(['meta', 'form'], action)
+      if (!equals(form, 'coinifyRecurringCheckout')) return
```
Comment (tony-blockchain) [inline packages/blockchain-wallet-v4-frontend/src/data/modules/coinify/sagas.js]: name of the form can be put to constant, in model.js file ( in the sagas, selectors, ... bundle)

File: packages/blockchain-wallet-v4-frontend/src/modals/CoinifyExchangeData/Payment/mediumHelpers.js [hunk 3]
```diff
@@ -187,6 +182,45 @@ export const bankOptionHelper = (
...
+      />
+    }
+    if (equals(bankDisabled, 'disable_kyc')) {
+      return (
+        <Fragment>
+          <FormattedMessage
+            id='scenes.buysell.coinifyexchangedata.payment.bank.unavailable_kyc'
+            defaultMessage='Bank transfers are unavailable until Identity Verification has been finished.'
+          />
+          {equals(path(['state'], kyc), 'pending') ? (
```
Comment (tony-blockchain) [inline packages/blockchain-wallet-v4-frontend/src/modals/CoinifyExchangeData/Payment/mediumHelpers.js]: `path` => `prop`

File: packages/blockchain-wallet-v4-frontend/src/modals/CoinifyExchangeData/Payment/template.success.js [hunk 4]
```diff
@@ -129,12 +132,17 @@ const Payment = props => {
...
   const cardDisabled = isCardDisabled(quote, limits)
-  const bankDisabled = isBankDisabled(quote, limits, kycState)
+  const bankDisabled = isBankDisabled(quote, limits, kycState, isSubscriptionTrade)
   if (bankDisabled && medium !== 'card') handlePaymentClick('card')
   const prefillCardMax = limits => handlePrefillCardMax(limits)
 
   const isChecked = type => medium === type
 
+  const onClickHelper = () => {
+    if (path(['name'], level) < 2 && medium === 'bank') triggerKyc()
```
Comment (tony-blockchain) [inline packages/blockchain-wallet-v4-frontend/src/modals/CoinifyExchangeData/Payment/template.success.js]: `path` => `prop`

File: packages/blockchain-wallet-v4-frontend/src/modals/CoinifyExchangeData/Payment/template.success.js [hunk 5]
```diff
@@ -178,34 +187,22 @@ const Payment = props => {
...
-                fullwidth
-                disabled={!medium || busy}
-              >
-                {busyHelper(busy)}
-              </StepTransition>
-            )}
+            <Button
+              nature='primary'
+              fullwidth
+              onClick={() => onClickHelper()}
```
Comment (tony-blockchain) [inline packages/blockchain-wallet-v4-frontend/src/modals/CoinifyExchangeData/Payment/template.success.js]: `() =>` usage of callback props recreates them on each render, it's better to use `onClick={onClickHelper}` if it's not becoming too bulky(e.g. binds)

File: packages/blockchain-wallet-v4-frontend/src/scenes/BuySell/CoinifyCheckout/Content/Buy/index.js [hunk 6]
```diff
@@ -62,33 +65,51 @@ class CoinifyBuyContainer extends React.Component {
...
+          subscriptions={subscriptions}
+          showModal={showModal}
+          showRecurringBuy={showRecurringBuy}
+          showRecurringModal={showRecurringModal}
+          subscription={subscription}
+          subscriptionData={subscriptionData}
+          trade={trade}
+          trades={trades}
+          value={value}
+          clearTradeError={() =>
```
Comment (tony-blockchain) [inline packages/blockchain-wallet-v4-frontend/src/scenes/BuySell/CoinifyCheckout/Content/Buy/index.js]: same thing with `() =>` for these props

File: packages/blockchain-wallet-v4-frontend/src/scenes/BuySell/CoinifyCheckout/Content/Buy/template.success.js [hunk 7]
```diff
@@ -70,78 +75,82 @@ const CoinifyBuy = props => {
...
+            setMax={setMax}
+            setMin={setMin}
+            showRecurringBuy={showRecurringBuy}
+            showRecurringModal={showRecurringModal}
+            symbol={symbol}
+            type={'buy'}
+          />
+        </LeftContainer>
+        <RightContainer>
+          {activeSubscriptions.length > 0 ? (
```
Comment (tony-blockchain) [inline packages/blockchain-wallet-v4-frontend/src/scenes/BuySell/CoinifyCheckout/Content/Buy/template.success.js]: `isEmpty` can be used here

File: packages/blockchain-wallet-v4-frontend/src/scenes/BuySell/CoinifyCheckout/Content/Buy/template.success.js [hunk 8]
```diff
@@ -70,78 +75,82 @@ const CoinifyBuy = props => {
...
+            <NextSubscription
+              subscriptions={subscriptions}
+              trades={trades.filter(t => t.tradeSubscriptionId)}
```
Comment (tony-blockchain) [inline packages/blockchain-wallet-v4-frontend/src/scenes/BuySell/CoinifyCheckout/Content/Buy/template.success.js]: `filter(prop('tradeSubscriptionId'))` could help avoid null pointer and remove 1-digit variable

File: packages/blockchain-wallet-v4-frontend/src/scenes/BuySell/CoinifyCheckout/Content/Buy/template.success.js [hunk 9]
```diff
@@ -70,78 +75,82 @@ const CoinifyBuy = props => {
...
-                busy={busy}
-                type={'buy'}
-                clearTradeError={clearTradeError}
-              />
-            </OrderSubmitWrapper>
-          </CheckoutWrapper>
-        </StepView>
-      </Stepper>
+          ) : null}
+          {path(['state'], kyc) ? (
```
Comment (tony-blockchain) [inline packages/blockchain-wallet-v4-frontend/src/scenes/BuySell/CoinifyCheckout/Content/Buy/template.success.js]: path => prop

File: packages/blockchain-wallet-v4-frontend/src/scenes/BuySell/CoinifyCheckout/Content/Buy/template.success.js [hunk 10]
```diff
@@ -70,78 +75,82 @@ const CoinifyBuy = props => {
...
+            <KYCNotification
+              kyc={kyc}
+              limits={limits.buy}
+              symbol={symbol}
+              onTrigger={kyc => handleKycAction(kyc)}
```
Comment (tony-blockchain) [inline packages/blockchain-wallet-v4-frontend/src/scenes/BuySell/CoinifyCheckout/Content/Buy/template.success.js]: `() =>` notation

File: packages/blockchain-wallet-v4-frontend/src/scenes/BuySell/CoinifyCheckout/Content/KYCNotification/template.js [hunk 11]
```diff
@@ -51,15 +51,15 @@ const KYCNotification = props => {
                   id='scenes.buysell.coinifycheckout.content.kycnotification.limitsnotice.sell'
                   defaultMessage='While your identity gets verified, you can sell up to '
                 />
-                {sellMax} BTC.
+                &nbsp;{sellMax} BTC.
```
Comment (tony-blockchain) [inline packages/blockchain-wallet-v4-frontend/src/scenes/BuySell/CoinifyCheckout/Content/KYCNotification/template.js]: default message has a space in the end

File: packages/blockchain-wallet-v4/src/redux/data/coinify/selectors.js [hunk 12]
```diff
@@ -28,6 +28,8 @@ export const getLimits = state => getProfile(state).map(path(['_limits']))
 
 export const getLevel = state => getProfile(state).map(path(['_level']))
 
+export const getTradeSubscriptionsAllowed = state => getProfile(state).map(path(['_tradeSubscriptionsAllowed']))
```
Comment (tony-blockchain) [inline packages/blockchain-wallet-v4/src/redux/data/coinify/selectors.js]: path => prop

File: packages/blockchain-wallet-v4-frontend/src/scenes/BuySell/CoinifyCheckout/Content/KYCNotification/template.js (hunk 11 above)
Comment (prwelber) [reply packages/blockchain-wallet-v4-frontend/src/scenes/BuySell/CoinifyCheckout/Content/KYCNotification/template.js]: Good eye, but it looks like the space in the message isn't being rendered. Keeping the `&nbsp;` and removing the space from the message so there's no confusion.

**Comment by prwelber** [comment]
Remaining Tasks

- [ ] fix tests
- [ ] remove `() =>` notation where necessary 
- [ ] add confirm modal to cancel trade

File: packages/blockchain-wallet-v4-frontend/src/data/modules/coinify/sagas.spec.js [hunk 13]
```diff
@@ -1267,4 +1269,131 @@ describe('coinifySagas', () => {
...
+      const action = {
+        payload: 'weekly',
+        meta: {
+          form: RECURRING_CHECKOUT_FORM,
+          field: 'frequency'
+        }
+      }
+      let saga = testSaga(handleRecurringFormChange, action)
+
+      it('should set show modal to true', () => {
```
Comment (tony-blockchain) [inline packages/blockchain-wallet-v4-frontend/src/data/modules/coinify/sagas.spec.js]: looks like a wring test description

File: packages/blockchain-wallet-v4-frontend/src/services/CoinifyService/index.spec.js [hunk 14]
```diff
@@ -210,8 +210,8 @@ describe('CoinifyService', () => {
     })
     it('should return the correct text for completed status', () => {
```
Comment (tony-blockchain) [inline packages/blockchain-wallet-v4-frontend/src/services/CoinifyService/index.spec.js]: I guess it should be 'should return no text ...'

**Comment by keonglee** [comment]
pull request

**Comment by BitcoinInvestors** [comment]
I propose for litecoin in this wallet babe

**Pull request closed without merging**